resources/js/components/Menu.vue

```diff
@@ -0,0 +1,148 @@
+            :is="item.type"
+            :options="item.options"
+            :key="index"
+          ></component>
+        </template>
+      </b-col>
+
+      <b-col class="text-right" v-if="sectionRight">
+        <template v-for="(item, index) in section.right">
+          <b-button-group v-if="isVisible(item, 'group')" size="sm" :key="index">
+            <b-button
+              v-for="(button, indexButton) in item.items"
+              :variant="button.variant || 'secondary'"
+              class="text-capitalize"
+              :title="button.title"
+              :key="`group-${indexButton}`"
+              @click="button.action"
+              v-if="button.hide !== true"
+            >
+              <i :class="button.icon"></i>
+              {{ button.name }}
+            </b-button>
+          </b-button-group>
+
+          <b-button
+            v-if="isVisible(item, 'button')"
+            size="sm"
+            :variant="item.variant || 'secondary'"
+            class="text-capitalize"
+            :title="item.title"
+            :key="index"
+            @click="item.action"
+          >
+            <i :class="item.icon"></i>
+            {{ item.name }}
+          </b-button>
+
+          <component
+            v-if="item.type !== 'group' && item.type !== 'button'"
+            :is="item.type"
+            :options="item.options"
+            :key="index"
+          ></component>
+        </template>
+      </b-col>
+    </b-row>
+  </b-card-header>
+</template>
+
+
+<script>
+export default {
+  props: ['options'],
+  data() {
+    return {
+      changeItems: {},
+      newItems: [],
+      sectionRight: true,
+      items: []
+    };
+  },
+  watch: {},
+  computed: {
+    section() {
+      let response = {};
+      response.left = [];
+      response.right = [];
+      this.items = [];
+
+      this.options.forEach(item => {
+        if (item.type === 'group') {
+          item.items.forEach(sub => {
+            this.items.push(sub.id);
+            return Object.assign(sub, this.changeItems[sub.id]);
+          });
+        }
+        response[item.section].push(
+          Object.assign(item, this.changeItems[item.id])
+        );
+        this.items.push(item.id);
+      });
+      this.newItems.forEach(item => {
+        if (this.items.indexOf(item.id) === -1) {
+          response[item.section].push(item);
+          this.items.push(item.id);
+        }
+      });
+      return response;
+    }
+  },
+  methods: {
+    isVisible(item, type) {
+      return item.type === type && item.hide !== true;
+    },
+    changeItem(id, value) {
+      this.changeItems[id] = value;
+      this.changeOptions(this.section.left);
+      this.changeOptions(this.section.right);
+    },
+    changeOptions(data) {
+      data.forEach(item => {
+        item = Object.assign(item, this.changeItems[item.id]);
+        if (item.type === 'group') {
+          this.changeOptions(item.items);
+        }
+      });
+    },
```

Comment:
These two methods are not called anywhere in this component. It seems they are set up to only be called externally, through a `ref`. This poses a problem, since these methods are part of this components internal state, and it will be assumed they won't be called from outside the component directly. This makes refactoring very difficult. But also makes understanding this code more difficult too, as now we have to jump between two components to fully understand how this is suppose to work.

From [the docs](https://vuejs.org/v2/guide/components-edge-cases.html#Accessing-Child-Component-Instances-amp-Child-Elements), regarding refs:
> It is only meant as an escape hatch for direct child manipulation

Please refactor this to use the component's _public_ API, that is: ↓**props down**↓, ↑**events up**↑. 

`section` is already a computed property, so you can delete these two methods, and wherever you need to change a group or button, change it in the options that are passed down through the `options` prop.

Reply:
yes,  "ref" was actually used to call these methods